ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/backwardsync/BackwardsSyncAlgorithm.java

```diff
@@ -0,0 +1,194 @@
+ *  * specific language governing permissions and limitations under the License.
+ *  *
+ *  * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+package org.hyperledger.besu.ethereum.eth.sync.backwardsync;
+
+import static org.hyperledger.besu.util.Slf4jLambdaHelper.debugLambda;
+import static org.slf4j.LoggerFactory.getLogger;
+
+import org.hyperledger.besu.datatypes.Hash;
+import org.hyperledger.besu.ethereum.chain.MutableBlockchain;
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+
+public class BackwardsSyncAlgorithm {
+  private static final Logger LOG = getLogger(BackwardsSyncAlgorithm.class);
+
+  private final BackwardSyncContext context;
+  private final FinalBlockConfirmation finalBlockConfirmation;
+  private volatile boolean finished = false;
+
+  public BackwardsSyncAlgorithm(
+      final BackwardSyncContext context, final FinalBlockConfirmation finalBlockConfirmation) {
+    this.context = context;
+    this.finalBlockConfirmation = finalBlockConfirmation;
+  }
+
+  public CompletableFuture<Void> executeBackwardsSync(final Void unused) {
+    final CompletableFuture<Void> nextStep = pickNextStep();
+    if (finished) {
+      return nextStep;
+    }
+    return nextStep.thenCompose(this::executeBackwardsSync);
+  }
+
+  public CompletableFuture<Void> pickNextStep() {
+    final Optional<Hash> firstHash = context.getBackwardChain().getFirstHashToAppend();
+    if (firstHash.isPresent()) {
+      return executeSyncStep(firstHash.get());
+    }
+    if (!context.isReady()) {
+      return waitForReady();
+    }
+    runFinalizedSuccessionRule(
+        context.getProtocolContext().getBlockchain(), context.findMaybeFinalized());
+    final Optional<BlockHeader> firstAncestorHeader =
+        context.getBackwardChain().getFirstAncestorHeader();
+    if (firstAncestorHeader.isEmpty()) {
+      this.finished = true;
+      LOG.info("The Backward sync is done...");
+      context.getBackwardChain().clear();
+      return CompletableFuture.completedFuture(null);
+    }
+    final MutableBlockchain blockchain = context.getProtocolContext().getBlockchain();
+    if (blockchain.contains(firstAncestorHeader.get().getHash())) {
+      return executeProcessKnownAncestors();
+    }
+    if (blockchain.getChainHead().getHeight() > firstAncestorHeader.get().getNumber()) {
+      debugLambda(
+          LOG,
+          "Backward reached bellow previous head {}({}) : {} ({})",
+          () -> blockchain.getChainHead().getHeight(),
+          () -> blockchain.getChainHead().getHash().toHexString(),
+          () -> firstAncestorHeader.get().getNumber(),
+          () -> firstAncestorHeader.get().getHash());
```

Comment:
suggestion: create and use a `ChainHead::toLogString` that return a string with the same format as `BlockHeader::toLogString` for log standardization